### PR TITLE
Add async processing for patient bulk uploads

### LINKED ISSUE #4371: Remove role restrictions on patient bulk upload
### Background
Patient bulk upload is currently only available to SimpleReport site admins (i.e., engineers). The final stage of opening the feature to end users will be changing the role restrictions to allow regular SimpleReport admins access.

### Action requested
Change the authorization configuration on `UploadService.processPersonCSV` from `RequireGlobalAdminUser` to `RequirePermissionEditFacility`. (Technically this is a patient-related operation, but given that it will affect many patients this permission level seems more appropriate).

https://github.com/CDCgov/prime-simplereport/blob/master/backend/src/main/java/gov/cdc/usds/simplereport/service/UploadService.java#L92-L92

### Acceptance Criteria
Bulk upload is available to SimpleReport admin users, not just site admins.

### Additional context
This is the last piece of work to do for patient bulk upload - everything else (UI updates, error messaging, etc) should be finished before this is merged.

### LINKED ISSUE #4662: Add batching and async for patient bulk uploader
### Background
We'd like the upload process to 1) provide faster status update to users and 2) be faster overall.

### Action requested
- First, we want to enable batch saving to the database using `saveAll()`.
- Second, we'd like the save to run in a separate thread so the backend can continue processing the save independent of the validation, so it's not tied to the response sent to the user for validation. This can likely be done using `@Async` or by setting up a single separate thread.

### Acceptance Criteria
User gets a quick response notifying them of successful/unsuccessful validation, and the save operation continues independently such that if the operation takes longer than 60 seconds the user does not get a time out error response.

### Additional context
If batch upload increases processing time satisfactorily to less than 60 seconds, adding async may not be necessary.

See spike #4553 for more context.

Here is an example of the error we don't want to see:
Screen Shot 2022-10-26 at 2 28 01 PM

### PR DESCRIPTION
**Pull request by emmastephenson**
# BACKEND PULL REQUEST

## Related Issue

- Fixes #4662 
- Also fixes #4371 

## Changes Proposed

- Adds async processing for patient saving in the CSV bulk uploader
- Expected behavior is that validation feedback is sent immediately, and a response is sent to the frontend while patients are still being saved
- Should not trigger Akamai in deployed environments

## Additional Information

There are two main changes required to get `async` working.

  ### 1. Change the security context to `INHERITABLETHREADLOCAL`. 

This means any new threads spawned will inherit the security context of the parent thread. This is required because we need the security context in `PatientBulkUploadServiceAsync` to verify that the user has permissions to call the method.
  - We were previously using `MODE_THREADLOCAL`, the default
  - [Spring documentation on `INHERITABLETHREADLOCAL`](https://docs.spring.io/spring-security/site/docs/3.0.x/reference/technical-overview.html#:~:text=Other%20applications%20might%20want%20to%20have%20threads%20spawned%20by%20the%20secure%20thread%20also%20assume%20the%20same%20security%20identity.%20This%20is%20achieved%20by%20using%20SecurityContextHolder.MODE_INHERITABLETHREADLOCAL)
  - [More useful random documentation on `INHERITABLETHREADLOCAL`](https://howtodoinjava.com/spring-security/spring-security-context-propagation/#:~:text=the%20parent%20thread.-,MODE_INHERITABLETHREADLOCAL,-%3A%20It%20copies%20the)

  ### 2. Add try/catch blocks around `getCurrentOrganizationRoles` and `getCurrentApiUser` for `ScopeNotActiveException`. 

This is the source of the dreaded "trying to access request outside of requesting thread" error. In this case it was actually true - we were trying to access `CurrentOrganizationRolesContextHolder` and `CurrentApiUserContextHolder` _after_ the initial request had closed. The stack trace on these goes something like the following:
  1. `PatientBulkUploadServiceAsync` calls `_personService.addPatientsAndPhoneNumbers`
  2. The `@Authorization` annotation on that function eventually resolves to [`userHasPermissions`](https://github.com/CDCgov/prime-simplereport/blob/7e04b7a00f54739ee9589a7abfb7248db8d297ca/backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/UserAuthorizationVerifier.java#L86-L86)
  3. That in turn calls the `getCurrentOrganizationRoles` and `getCurrentApiUser` methods
  4. Without the try/catch, the backend blows up because there's no longer a request in scope

In my opinion, this is not a security risk because the fallback methods we use to verify permissions ([`getCurrentApiUserNoCache`](https://github.com/CDCgov/prime-simplereport/blob/7e04b7a00f54739ee9589a7abfb7248db8d297ca/backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java#L520-L520) and [`fetchCurrentOrganizationRoles`](https://github.com/CDCgov/prime-simplereport/blob/7e04b7a00f54739ee9589a7abfb7248db8d297ca/backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java#L67-L67) ) in turn rely on the [`LoggedInAuthorizationService`](https://github.com/CDCgov/prime-simplereport/blob/7e04b7a00f54739ee9589a7abfb7248db8d297ca/backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java#L34-L34) to determine roles and permissions. That service relies on `SecurityContextHolder` to determine user permissions, which is populated thanks to our nifty `MODE_INHERITABLETHREADLOCAL` above. So, if the request is out of scope but `SecurityContextHolder` is empty, an exception will eventually be thrown - just later down the stack.

TL;DR the ContextHolder pattern is a bit of a cheat to store user permissions inside a request scope, but now that we're operating outside a request scope they don't work. We can get around that to still access user permissions though.
 

## Testing

- Bulk uploading patients returns an immediate response for validation, and patients are eventually loaded into the database
- All other security / context pieces of the application still WAI

## EDIT: Testing Setup
Backend tests for this were failing for a while on seemingly unrelated tests (`SmsServiceTest` and `TestResultUploadTest`). This only happened when the full test suite was run. After in-depth investigation, I can confidently say that this was unrelated to the async changes in this PR and simply brought on as a result of adding another `SpringBootTest` to our setup - the database ran out of connections with so many tests. (@rin-skylight and I did thoroughly debug the code and ensured that it was properly opening and closing database connections in the new threads).
The solution was to add an alteration to the `reset-db` script, to increase the number of available connections from 100 to 500. This fixed the issue when running the test suite both locally and in CI.

## Checklist for Primary Reviewer
- [x] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [ ] Changes comply with the SimpleReport Style Guide
- [ ] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)

### CODE REVIEW

**Comment by georgehager** [comment]
@emmastephenson I am going to play around with adding `CompletableFuture` onto this implementation, similar to what [Boban experimented with here](https://github.com/CDCgov/prime-simplereport/commit/3a76e1636ec54602ad607aac8ad9abb7382a781c).

This may speed up our performance even more.

**Comment by georgehager** [comment]
@emmastephenson we may want to rebase this on my branch george/4662-batch-save-patients, or we could wait until that's merged into main. I just tried it and there are some little conflicts.

**Comment by emmastephenson** [comment]
@georgehager yeah, I started rebasing and then figured I would just wait until it's merged into `main` 😆 I also have to fix all the tests before this can be reviewed/merged.

**Comment by emmastephenson** [comment]
Dropping this here so we have the reference for later: https://github.com/spring-projects/spring-security/issues/6856

what I'm understanding from that is that we need to make sure our threads aren't pooled, because the INHERITABLETHREADLOCAL will run into issues if using thread pools. Not sure what the default SpringBoot setup is.

**Comment by georgehager** [comment]
> Dropping this here so we have the reference for later: [spring-projects/spring-security#6856](https://github.com/spring-projects/spring-security/issues/6856)
> 
> what I'm understanding from that is that we need to make sure our threads aren't pooled, because the INHERITABLETHREADLOCAL will run into issues if using thread pools. Not sure what the default SpringBoot setup is using.

@emmastephenson I had this same concern too when reading through some docs, also not sure what the default is.

**Comment by emmastephenson** [comment]
[Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/scheduling/annotation/EnableAsync.html) for the `@EnableAsync` annotation would indicate that if no `Executor` bean is provided, it falls back to using a [`SimpleAsyncTaskExecutor`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/task/SimpleAsyncTaskExecutor.html). Javadoc for that class specifically says that:

> [NOTE: This implementation does not reuse threads! Consider a thread-pooling TaskExecutor implementation instead, in particular for executing a large number of short-lived tasks.](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/core/task/SimpleAsyncTaskExecutor.html#:~:text=NOTE%3A%20This%20implementation%20does%20not%20reuse%20threads!%20Consider%20a%20thread%2Dpooling%20TaskExecutor%20implementation%20instead%2C%20in%20particular%20for%20executing%20a%20large%20number%20of%20short%2Dlived%20tasks.)

So, since we're apparently not using a thread-pooling application, I believe we're ok to continue using MODE_INHERITABLETHREADLOCAL. But definitely something to watch out for if we ever want to use thread pooling.

**Comment by emmastephenson** [comment]
Also note that if you're like me and paranoid that you've seen something about pooling in the application before: we do have specific pool [properties set](https://github.com/CDCgov/prime-simplereport/blob/3769a8d82959f99622be1875d7798a8cda6b2050/backend/src/main/resources/application.yaml#L12-L12) for our datasource/hikari connections, but that's unrelated to thread pooling; it's just the number of database connections the app will maintain at once.

**Comment by georgehager** [comment]
> 

@emmastephenson I did another search for any implementation we might have of task executor/thread pool, and the only thing we reference is `ThreadPoolTaskScheduler`, but only in the limited context `ScheduledTasksService.java` and `ReminderService.java`. I agree I'm certain we're falling back to the default `SimpleAsyncTaskExecutor`. Another piece of information I found relevant is this, from the [Spring SecurityContextHolder docs](https://docs.spring.io/spring-security/site/docs/5.2.0.M4/reference/htmlsingle/#securitycontextholder-securitycontext-and-authentication-objects):

> By default the `SecurityContextHolder` uses a `ThreadLocal` to store these details, which means that the security context is always available to methods in the same thread of execution, even if the security context is not explicitly passed around as an argument to those methods. Using a `ThreadLocal` in this way is quite safe if care is taken to clear the thread after the present principal’s request is processed. Of course, Spring Security takes care of this for you automatically so there is no need to worry about it.

**Comment by emmastephenson** [comment]
Note: tests are still failing atrociously, but I wanted to push the work up anyways.

Now instead of failing because the user isn't authenticated, the tests are failing because there's an n+1 call to `DeviceSpecimenType` somewhere in the stack. Suppressing the n+1 exception doesn't help, because the tests continue to run in an infinite loop until the `awaitility` library times out. This doesn't happen in any other test environment AFAIK, so it definitely seems related to the thread changes 🤔

**Comment by emmastephenson** [comment]
Fun thing happening with this PR - we appear to be exceeding some timeout limits on our testing setup. `SmsServiceTest` and `TestResultUploadServiceTest` consistently fail when run as part of `gradle test` or the CI suite, but pass when run by themselves. My theory is that these are the last two tests to run, and the `test db` docker database either times out or is killed by the time we get to these tests. Looking into resolution for that, but the rest of this PR should be good to review.

**Review by emyl3 — COMMENTED** [review]
one question for you emma! thank you for all your work on this 🙌

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java:518]: If I remember correctly, `getCurrentApiUserNoCache` is implemented here for tests, correct?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:63]: Can you remind me why we don't need to `organizationRolesContext.setOrganizationRoles(result);` here?

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:33]: is this authorization change premature here? my understanding is the feature will be available to org or facility managers, which this change would turn on, right? Do we want to wait to update this or go ahead?

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java
Comment (georgehager) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java:32]: I wonder if we should consider renaming these base tests at some point given this async service test is not at all related to graphql.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java:518]: In this instance, I believe it's also required for production. (You may be thinking of the changes to `DemoOktaRepository` and the `tenantDataAccessContextHolder` - those are specifically used for tests/non-Okta)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:63]: Yep! `organizationRolesContext` at this point no longer exists - it only exists while the scope of the request is active, and at this point in our async method we no longer have a request scope to work with. So trying to `setOrganizationRoles` will result in a `NullPointerException`.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:33]: Good catch - it is a little premature, and this will indeed open up the endpoint to org/facility managers. The reason I made the change is because having the existing permissions made testing nearly impossible - you needed a user that had global permissions to add the patients, but then to verify the patients were added you needed a user with permission to view the patients at a facility. Rather than try to work that permissioning out I just made this change, since it was next up in the work stream regardless. There's still no UI available for it (that little checkbox on the frontend does check for superuser permissions before displaying), so org/facility users shouldn't notice a change. They would be able to hit the endpoint directly with a CURL request, but I still think that's a low-risk enough activity to allow for now.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java:32]: I think the correct solution here might be for me to redefine all the useful bits of `BaseGraphQlTest` into a different base test 😄 I'm mostly using it for the setup/teardown properties and authentication it offers, but it does have some graphql-specific methods that I'm just ignoring. Then we could have something like

```
BaseAuthenticatedFullStackTest extends BaseFullStackTest
BaseGraphQlTest extends BaseAuthenticatedFullStackTest
PatientBulkUploadServiceAsyncTest extends BaseAuthenticatedFullStackTest
```
I'm going to wait to do that until the test suite is passing consistently, though 😄 (We're almost there!! It now passes the full suite locally but not in CI)

**Review by rin-skylight — APPROVED** [review]
I have no security concerns! After spending eight hours with this code, I can safely say that it looks great to my eyes. 😆

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (georgehager) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:33]: Yeah I thought I remembered this was necessary for tests. I'm fine with it, we may just be able to close the other ticket out at the same time.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java
Comment (georgehager) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java:32]: That sounds like a good separate ticket 😜

**Review by georgehager — DISMISSED** [review]
Looks like the tests are finally passing - thank you for your help @rin-skylight! 🎉 

I'm good with this, albeit probably missing something after looking at the code a lot over the last couple weeks 😅  I think there's some opportunity for cleanup with the test setup, but for the sake of time and and QA I'd be satisfied with having that pushed to a followup ticket.

File: backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java
Comment (emmastephenson) [reply backend/src/test/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsyncTest.java:32]: Haha good call! Here it is: https://github.com/CDCgov/prime-simplereport/issues/4848

**Comment by emmastephenson** [comment]
@emyl3 @BobanL @zdeveloper this PR is finally ready for a final review! 😭 It's deploying out to dev4, and you can use the [sample CSV in the secret repo](https://github.com/cdcent/simplereport_docs/blob/main/runbooks/csv_operations/patient_upload_final.csv) to test with if you don't have one already. 

@zdeveloper - we're also going to bug bash on the patient bulk upload before opening it fully to users, but that bash will incorporate the frontend and email changes as well.

**Review by BobanL — APPROVED** [review]
Everything looks good to me! Thank a bunch 👍

File: .github/workflows/test.yml
Comment (zdeveloper) [inline .github/workflows/test.yml:64]: whats the point of this docker cmd? I can see it restarts the sr_test_db container, not sure why, also I thought we werent using docker for the db in ci

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:23]: idk why this comment exists, the log is in git.

File: .github/workflows/test.yml
Comment (rin-skylight) [reply .github/workflows/test.yml:64]: A restart is necessary due to the change made in `reset-db.sql`, which increases the number of connections allocated to the database. That setting does not take effect until the database is forcibly restarted.

We are using docker for the database in the testing flow.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:31]: the underscore before local variable names violates the [java naming conventions](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html#:~:text=Except%20for%20variables%2C%20all%20instance%2C%20class%2C%20and%20class%20constants%20are%20in%20mixed%20case%20with%20a%20lowercase%20first%20letter.%20Internal%20words%20start%20with%20capital%20letters.%20Variable%20names%20should%20not%20start%20with%20underscore%20_%20or%20dollar%20sign%20%24%20characters%2C%20even%20though%20both%20are%20allowed.)

**Review by zdeveloper — CHANGES_REQUESTED** [review]
LGTM! mostly, however, this code is error prone, given a big enough file, you can potentially run out memory, or overwhelm the db with a very big saveAll list. I recommend chucking the processing and saveAll to around a 1000 patients at a time, that way you have a max memory limit.

File: .github/workflows/test.yml
Comment (zdeveloper) [reply .github/workflows/test.yml:64]: In that case, it makes more sense to me to just start the container with the correct [max-connections set](https://hub.docker.com/_/postgres#:~:text=postgres%20%2Dc%20shared_buffers%3D256MB%20%2Dc%20max_connections%3D200), 

If the github actions magical syntax doesn't support the passing of arguments, can we just add a separate step for running the docker command directly at the beginning of the steps?

`docker run -d -p 5432:5432 --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5 -e POSTGRES_PASSWORD=this_is_a_super_secure_admin_password postgres:13-alpine3.16 -c max_connections=500`

Comment (emmastephenson) [reply .github/workflows/test.yml:64]: The benefit of doing it this way is that we could set the max value in the `reset-db` script (which `create-db.sh` uses) and that max value is used both here and in the docker test setup (used for local testing). If we change the docker run command directly here, we'd still need to change the value for local docker-compose. There was a prior commit that used that approach (back when we were testing locally and hadn't yet figured out that CI is different), [4a1ef269](https://github.com/CDCgov/prime-simplereport/pull/4722/commits/4a1ef269a6ea4f6b085c60056898add9ca905656), but I prefer this way because it keeps the local test setup and the CI setup more in alignment.

The real solution would be to use the full docker-compose script in CI instead of this custom action, but that's out of scope for today 😅

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:23]: Will remove in a subsequent PR (I was just following convention from some of our other service classes, we're inconsistent 🤷)

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:31]: But what if I disagree with the style guides? 🤣 

JK, will remove in the subsequent PR. (I am curious why this style fell out of favor though, I really like keeping class-level fields and in-method variables distinguished with the underscore)

**Comment by emmastephenson** [comment]
@zdeveloper good call out on the large saveAll case, that does open us up to a bit of risk. We'll get that in to this PR ASAP but I'm going to save the other feedback for a subsequent PR I have planned as this one is already blocking other reviews. That followup PR should be in by EOW!

**Comment by georgehager** [comment]
@zdeveloper @emmastephenson I just pushed a commit with basic Spring level batching enabled. It's currently set to 100, because [multiple sources recommend 50-100 batch size, see this example](https://www.baeldung.com/spring-jdbc-batch-inserts#3-preparing-batch-inserts). This batch size should not negatively affect any of the existing `saveAll`s in our code - if they're small they effectively won't get batched so no performance degradation.

Besides this Spring level batching, we can add a layer of batching in the service itself, [similar to this](https://persistencelayer.wixsite.com/springboot-hibernate/post/the-best-way-to-batch-inserts-via-saveall-iterable-s-entities). I think Spring level batching is sufficient for now, we could add this on in a followup.

Note that if you uncomment `generate_statistics=true` in `application.yml` you can observe the effect of turning on batching.

File: .github/workflows/test.yml
Comment (zdeveloper) [reply .github/workflows/test.yml:64]: I agree about the docker-compose solution, can you please write a ticket to document and implement that solution, the current solution is certainly a half measure

**Review by rin-skylight — DISMISSED** [review]

**Review by georgehager — APPROVED** [review]
send it 🚀

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (zdeveloper) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:31]: Yeah, its a convention in c++, its kinda nice, but its not really necessary in java convention since it pushes the getter/setter accessor patterns and you shouldn't have public local variables, because of that you end up underscoring almost everything anyway, which is not really that useful.

File: .github/workflows/test.yml
Comment (emmastephenson) [reply .github/workflows/test.yml:64]: Done: https://github.com/CDCgov/prime-simplereport/issues/4856

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:23]: Ticket created to capture this and other followup work: https://github.com/CDCgov/prime-simplereport/issues/4857

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:31]: That's true, I didn't think about the possibility of it encouraging public local variables. Created this ticket to make sure it gets addressed: https://github.com/CDCgov/prime-simplereport/issues/4857

Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadService.java:31]: And it's even in the wiki now: https://github.com/CDCgov/prime-simplereport/wiki/SimpleReport-Style-Guide#variable-names

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:54]: this Set is still error prune, and susceptible to and out-of-memory error given a big enough file, since it will grow to the max number of the uploaded file.

**Review by zdeveloper — CHANGES_REQUESTED** [review]
Thanks @georgehager for adding batching on the saveAll, however, there still an outstanding issue to be addressed.

File: backend/src/main/resources/application.yaml
Comment (zdeveloper) [inline backend/src/main/resources/application.yaml:32]: @georgehager do you know whats the default setting for this feature? I am a bit concerned about any potential side effects

Comment (zdeveloper) [reply backend/src/main/resources/application.yaml:32]: I would recommend implementing a naive chunking solution on the service side since that will address the batching on the jdbc side and also address the potential OOM error because the Set size, rather than turning this setting on for everything.

Comment (georgehager) [reply backend/src/main/resources/application.yaml:32]: After a bit of reading, my opinion is we should do the naive chunking as you suggest in batches of 1000, which would pass at most 1000 records to `saveAll`, _and_ we should turn on `jdbc.batch_size` as well, without `order_inserts`. This would give us good memory safety and improved performance. I don't believe we should see side effects from just turning on `jdbc.batch_size` but I'll do some testing. @emmastephenson

**Comment by emmastephenson** [comment]
@zdeveloper can you give this another review when you get the chance? I just added batching like we talked about. TIA!

**Comment by emmastephenson** [comment]
Let the record show that the `dev5` deployment failure is unrelated to these changes 😅 It was caused by our cert issues this morning.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:45]: injecting this as a `@Value` is admittedly a little bit overkill, but it was the only reasonable way I could find to have separate values for `test` and actual development. (Otherwise testing this required a CSV size of 1001, which is just too large for unit testing)

File: backend/src/main/resources/application.yaml
Comment (emmastephenson) [reply backend/src/main/resources/application.yaml]: Long-term, I agree with @georgehager that including both naive chunking and batching the JDBC saves will give us the best performance. In the interest of keeping this PR small, however, I've created a new ticket to address the JDBC batching: https://github.com/CDCgov/prime-simplereport/issues/4864

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:160]: Is it ok that the batched calls to `addPatientsAndPhoneNumbers` are within a try block, but this one is not?

**Review by georgehager — APPROVED** [review]
lgtm! thanks for updating the tests for batching too. I had a comment about the final call to `addPatientsAndPhoneNumbers` not being in a try block but I'm not too worried about that because my email PR will wrap it anyway.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:160]: I think it is, because that method is not currently configured to throw any exceptions anyway - I think it would just silently fail.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (georgehager) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java:126]: did you mean to leave this comment in?

**Review by georgehager — DISMISSED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java
Comment (emmastephenson) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/PatientBulkUploadServiceAsync.java]: I certainly did not 😆 Thanks for catching that! Removed

**Review by georgehager — DISMISSED** [review]

**Comment by georgehager** [comment]
@zdeveloper when you get a chance, can you re-review this now that the naive batching changes are in? I'm pushing this PR forward per Emma's instructions while she's out.

**Review by rin-skylight — APPROVED** [review]
🚢

**Comment by georgehager** [comment]
Sorry @rin-skylight! Promise this is the last review request, just addressed some feedback from Zedd.

**Review by rin-skylight — APPROVED** [review]
🛳️ Bringing in the fleet of ships... 😆

**Review by zdeveloper — APPROVED** [review]
LGTM! thank you for all your hard work on this! Its gonna save us so much time!
This PR however will **break** local testing without okta. As you know, testing a patients file is done locally for PHI reasons.

@emmastephenson @georgehager Please make a ticket to address the changes needed for `DemoOktaRepository` to get this working locally without okta since many people on the team run their local dev without okta.

**Review by emyl3 — APPROVED** [review]
LGTM! Tested locally and on dev7. Thank you for all your work on this emma and george!!! 😹 

@georgehager  I also created  the ticket for the request Zedd made above: https://app.zenhub.com/workspaces/simplereport-growth--engagement-605b43a6ceb92c000f86279a/issues/gh/cdcgov/prime-simplereport/4900

Feel free to use this or discard/edit as needed!!

**Pull request merged**
